### ISSUE TITLE
DefaultMetricsRouteNameResolver fails for API projects

### ISSUE THREAD

**Issue by fatagun** [AppMetrics/AspNetCore#44]
[DefaultMetricsRouteNameResolver ](https://github.com/AppMetrics/AspNetCore/blob/87dcdc89ae25fdec868b9abf25b8ea46b762eaa4/src/App.Metrics.AspNetCore.Routing/DefaultMetricsRouteNameResolver.cs) returns empty string all the time for WebApi projects because of `Route`. 

When using [MvcRouteTemplateResolver  ](https://github.com/AppMetrics/AspNetCore/blob/dev/src/App.Metrics.AspNetCore.Mvc.Core/MvcRouteTemplateResolver.cs) with DefaultMetricsRouteNameResolver routes are not found. 

Registering DefaultMvcRouteTemplate resolver as follows:
services.AddMvc(options => options.Filters.Add(new MetricsResourceFilter(new MvcRouteTemplateResolver())));

I never get any data per endpoint.

**Comment by SukharevAndrey** [comment]
Same, it just do nothing on web api project

**Comment by fatagun** [comment]
You will need to create a new RouteTemplateResolver for API. Also, there are multiple ways of doing versioning.
ie: 
/api/v1/foo
/api/foo?api-version=v1.0
via headers etc.

Current impl only supports `[version](https://github.com/AppMetrics/AspNetCore/blob/8548650be44995eeb8020102f26c9bd30ded6ef8/src/App.Metrics.AspNetCore.Mvc.Core/MvcRouteTemplateResolver.cs#L20)`

**Comment by alhardy** [comment]
@fatagun yeah I've only supported at I needed re mulitple ways to support versioning. Would be awesome if you could help extend that?

@JoJo2406 thanks for the info, will have a look over the next few days

**Comment by fatagun** [comment]
@alhardy API versions regardless of the model (query string, route, header, etc) can be read via  [IApiVersionReader](https://github.com/Microsoft/aspnet-api-versioning/wiki/API-Version-Reader) provided by [Microsoft.AspNetCore.Mvc.Versioning](https://www.nuget.org/packages/microsoft.aspnetcore.mvc.versioning/) . This will introduce a new dependency. 

```
var apiversionreader = context.HttpContext.RequestServices.GetService<IApiVersionReader>();
var version = apiversionreader.Read(context.HttpContext.Request);
```
`HttpContext` is not available in `MvcRouteTemplateResolver` though. 

`RouteData` does not provide Api Version info, but `ActionDescriptor` has the information and can be retrieved as following in [MvcRouteTemplateResolver](https://github.com/AppMetrics/AspNetCore/blob/8548650be44995eeb8020102f26c9bd30ded6ef8/src/App.Metrics.AspNetCore.Mvc.Core/MvcRouteTemplateResolver.cs#L72) :

```
var controllerModel = actionDescriptor?.Properties[typeof(ControllerModel)] as ControllerModel;
var apiversionmodel = controllerModel?.Properties[typeof(ApiVersionModel)] as ApiVersionModel;
```

On the other hand, there is no `routeTemplate` if api versioning is done via headers, query string or media type. Not sure, how you would like to add the version info for these cases.

let me know if I and how I can help.

**Comment by alhardy** [comment]
@fatagun always open for help, no need to ask :)

**Comment by alhardy** [comment]
@otto-gebb There are updates made to asp.net core routing that has broken this. Since you're looking at it an alternate approach to a workaround could be to debug it and provide a PR if you have time.

I'm definitively going to look at it at some stage myself though if you can wait.

**Comment by otto-gebb** [comment]
The workaround I currently have:

```
services
    .AddMvc(opt => opt.EnableEndpointRouting = false)
    .SetCompatibilityVersion(CompatibilityVersion.Version_2_2)
```
